File: implementations/python/registry.py
import inspect
import json
from typing import Any, Callable, Dict

from pydantic import TypeAdapter

class ToolRegistry:
    def __init__(self):
        self._tools: Dict[str, Callable] = {}
        self._schemas: Dict[str, dict] = {}

    def register(self, name: str, description: str, func: Callable, schema: dict):
        """Register a tool with its JSON schema."""
        self._tools[name] = func
        self._schemas[name] = {
            "type": "function",
            "function": {
                "name": name,
                "description": description,
                "parameters": schema
            }
        }

    def get_tool_schemas(self) -> list[dict]:
        """Return the list of tool schemas for the LLM."""
        return list(self._schemas.values())

    def execute_tool(self, name: str, kwargs: dict) -> str:
        """Execute a tool by name and return its result as a string."""
        if name not in self._tools:
            return f"Error: Tool '{name}' not found."
        
        try:
            func = self._tools[name]
            result = func(**kwargs)
            
            # Ensure the result is always a string representation
            if not isinstance(result, str):
                result = json.dumps(result, default=str)
                
            return result
        except Exception as e:
            return f"Error executing tool '{name}': {str(e)}"

# A global registry instance for convenience
registry = ToolRegistry()

File: implementations/go/agent/registry.go
package agent

import (
	"encoding/json"
	"fmt"
	"github.com/sashabaranov/go-openai"
)

type ToolFunc func(args map[string]interface{}) (string, error)

type ToolRegistry struct {
	tools   map[string]ToolFunc
	schemas []openai.Tool
}

func NewToolRegistry() *ToolRegistry {
	return &ToolRegistry{
		tools:   make(map[string]ToolFunc),
		schemas: make([]openai.Tool, 0),
	}
}

func (r *ToolRegistry) Register(name, description string, f ToolFunc, parametersJson json.RawMessage) {
	r.tools[name] = f

	// Unmarshal the parametersJson into a map to pass to the library
	var parameters interface{}
	_ = json.Unmarshal(parametersJson, &parameters)

	tool := openai.Tool{
		Type: openai.ToolTypeFunction,
		Function: &openai.FunctionDefinition{
			Name:        name,
			Description: description,
			Parameters:  parameters,
		},
	}
	r.schemas = append(r.schemas, tool)
}

func (r *ToolRegistry) GetToolSchemas() []openai.Tool {
	return r.schemas
}

func (r *ToolRegistry) ExecuteTool(name string, argsJson string) string {
	f, ok := r.tools[name]
	if !ok {
		return fmt.Sprintf("Error: Tool '%s' not found.", name)
	}

	var args map[string]interface{}
	if err := json.Unmarshal([]byte(argsJson), &args); err != nil {
		return fmt.Sprintf("Error parsing arguments for tool '%s': %v", name, err)
	}

	result, err := f(args)
	if err != nil {
		return fmt.Sprintf("Error executing tool '%s': %v", name, err)
	}
	return result
}
